### PR TITLE
Fixes issue #17 (bug fixes & enhancements)

### PR DESCRIPTION
**Pull request by NamitS27**
### Fixes issue #17 
- Fixed `cd` and `date`(pretty print flag) command.
- Added new shell commands,
  - `ps` 
  - `mathexpr`
  - `rmdir`
  - `pwd`

### CODE REVIEW

**Comment by fruhland** [comment]
Thank you for your submission. We will review your code over the next few days and give you feedback, if anything needs to be updated.

**Review by fruhland — CHANGES_REQUESTED** [review]
Please remove all build artifacts from your Pull Request (e.g. `.d`, `.bin` files and the `intird`).

**Review by fruhland — COMMENTED** [review]
Please do not remove license headers and be sure to add them to all your new files. Feel free to add yourself as authors if you edited the regarding file, but do not remove the license header.

**Review by fruhland — CHANGES_REQUESTED** [review]
Please do not remove license headers (e.g. `DiskInfo`) and be sure to add them to all your new files. Feel free to add yourself as authors if you edited the regarding file, but do not remove the license header.

**Review by fruhland — CHANGES_REQUESTED** [review]
Removed...

**Review by fruhland — CHANGES_REQUESTED** [review]
Scheduler: Please do not expose `readyQueues` as a public variable. If you need to access it, implement a getter-method.

**Review by fruhland — CHANGES_REQUESTED** [review]
As far as we know, executing `cd` without arguments on a Unix-based system, does not do anything. That is why, we implemented the command this way. We think, this is a more natural behaviour than changing to the root directory.

File: src/application/shell/command/DiskInfo.cpp [hunk 1]
```diff
@@ -1,42 +1,31 @@
...
- * later version.
- *
- * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
- * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
- * details.
- *
- * You should have received a copy of the GNU General Public License
- * along with this program.  If not, see <http://www.gnu.org/licenses/>
- */
-
```
Comment (fruhland) [inline src/application/shell/command/DiskInfo.cpp:16]: Please do not remove license headers and be sure to add them to all your new files. Feel free to add yourself as authors if you edited the regarding file, but do not remove the license header.

File: src/application/shell/command/DiskInfo.h [hunk 2]
```diff
@@ -1,18 +1,3 @@
...
- * License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any
- * later version.
- *
- * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
- * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
- * details.
- *
- * You should have received a copy of the GNU General Public License
- * along with this program.  If not, see <http://www.gnu.org/licenses/>
- */
```
Comment (fruhland) [inline src/application/shell/command/DiskInfo.h:15]: Please do not remove license headers and be sure to add them to all your new files. Feel free to add yourself as authors if you edited the regarding file, but do not remove the license header.

File: src/application/shell/command/Ps.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        return;
+    }
+
+    /*Creates an instance of scheduler to access all the processes and counts the number of active processes*/
+    printf("Total Processes : %d\n",Kernel::Scheduler::getInstance().getThreadCount());
+    printf("-------------------------------\n");
+    printf("PID   Process Name\n");
+    printf("-------------------------------\n");
+    /*For each ready queue(based on priority order) we get the process info of all the processes in the queue*/
+    for (const auto &queue : Kernel::Scheduler::getInstance().readyQueues) {
```
Comment (fruhland) [inline src/application/shell/command/Ps.cpp:26]: Please do not expose readyQueues as a public variable. If you need to access it, implement a getter-method.

File: src/device/input/Keyboard.cpp [hunk 4]
```diff
@@ -26,18 +26,18 @@
 
 uint8_t Keyboard::normal_tab[] = {
         0, 0, '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', 225, 39, '\b',
-        0, 'q', 'w', 'e', 'r', 't', 'z', 'u', 'i', 'o', 'p', 129, '+', '\n',
+        0, 'q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p', 129, '+', '\n',
```
Comment (fruhland) [inline src/device/input/Keyboard.cpp:29]: Please do not change the keyboard layout. Feel free to implement multiple layouts and a switching mechanism (this is a planned feature, anyways).

File: src/kernel/thread/Scheduler.h [hunk 5]
```diff
@@ -138,7 +140,7 @@ class Scheduler {
 
     PriorityPattern &priority;
 
-    Util::Array<Util::ArrayBlockingQueue<Thread *>> readyQueues;
```
Comment (fruhland) [inline src/kernel/thread/Scheduler.h:141]: Please do not expose readyQueues as a public variable. If you need to access it, implement a getter-method.

File: src/application/shell/command/Cd.cpp [hunk 6]
```diff
@@ -30,6 +30,8 @@ void Cd::execute(Util::Array<String> &args) {
     }
 
     if(parser.getUnnamedArguments().length() < 1) {
+        /* For cd command with no arguments will change the current working directory to the root directory */
+        shell.setCurrentWorkingDirectory(Directory::open("/"));
```
Comment (fruhland) [inline src/application/shell/command/Cd.cpp:34]: As far as we know, executing cd without arguments on a Unix-based system, does not do anything. That is why, we implemented the command this way. We think, this is a more natural behaviour than changing to the root directory.

File: src/lib/util/ArgumentParser.cpp [hunk 7]
```diff
@@ -32,6 +32,13 @@ void ArgumentParser::addParameter(const String &name, const String &abbreviation
     }
 }
 
+/* Displays the whole command entered by the user in the shell */
+String ArgumentParser::getEnteredCommand(){
+    String command = "";
+    for(auto it:unnamedArguments) command += it + " ";
```
Comment (fruhland) [inline src/lib/util/ArgumentParser.cpp:38]: We think that single line loops are not very readable and would prefer use of brackets and blocks.

**Review by fruhland — CHANGES_REQUESTED** [review]

**Comment by NamitS27** [comment]
@fruhland, the changes suggested by you has been done. Can you please verify it & suggest any further changes if required.

File: src/application/shell/command/Mathexpr.cpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    Util::ArgumentParser parser(getHelpText(), 1);
+
+    /* Parser checks for the error if any on the basis of the command entered by the user */
+    if(!parser.parse(args)) {
+        stderr << args[0] << ": " << parser.getErrorString() << endl;
+        return;
+    }
+    /* Check whether more than one unamed arguments are given by the user and display the appropriate message it they do */
+    if(parser.getUnnamedArguments().length()>1){
+        stderr << args[0] << ": Invalid command" << endl;
```
Comment (fruhland) [inline src/application/shell/command/Mathexpr.cpp:142]: Could you please give a more detailed error message (e.g. "Invalid number of arguments")?

File: src/application/shell/command/Mathexpr.cpp [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+void Mathexpr::execute(Util::Array<String> &args) {
+    Util::ArgumentParser parser(getHelpText(), 1);
+
+    /* Parser checks for the error if any on the basis of the command entered by the user */
+    if(!parser.parse(args)) {
+        stderr << args[0] << ": " << parser.getErrorString() << endl;
+        return;
+    }
+    /* Check whether more than one unamed arguments are given by the user and display the appropriate message it they do */
```
Comment (fruhland) [inline src/application/shell/command/Mathexpr.cpp:140]: Please also check for zero arguments. Right now, the OS crashes when you execute "mathexpr" without any arguments.

**Review by fruhland — COMMENTED** [review]
Thanks for your update. Everything looks good now, however we found a small bug in "mathexpr": If you execute it without any arguments, the OS crashes. Can please have a look into that, before we merge your PR?

Best regards!

**Comment by NamitS27** [comment]
@fruhland I have fixed the mathexpr command for zero arguments and also edited the error message by making it more precise.

**Comment by fruhland** [comment]
Thanks! Everything is fine now :)

**Pull request merged**